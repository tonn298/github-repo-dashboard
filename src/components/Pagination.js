import React from "react";
import styled from "styled-components";

import Button from "../elements/Button";

const PaginationStyled = styled.div`
  display: flex;
  justify-content: center;
  margin-top: 20px;
  select {
    background-color: transparent;
    margin: 0;
    width: 220px;
    font-size: 20px;
    line-height: inherit;
    border: 1px solid #aaa;
    padding: 15px;
  }
`;

const Pagination = ({ paginate, currentPage, pagesArray }) => {
  const movePageBy = (number) => {
    const gotoPage = Number(currentPage) + number;
    if (gotoPage <= 0 || gotoPage > pagesArray.length) {
      return;
    }

    paginate(gotoPage);
    document.getElementById("page").value = gotoPage;
  };

  return (
    <PaginationStyled data-testid="pagination">
      <Button text="<<" handleOnClick={() => movePageBy(-2)} />
      <Button text="<" handleOnClick={() => movePageBy(-1)} />
      <select
        className="page-dropdown"
        name="page"
        id="page"
        onChange={(e) => paginate(e.target.value)}
      >
        {pagesArray.map((number, id) => {
          return (
            <option key={id} data-testid={`option-${id}`} value={number}>
              Current page: {number}
            </option>
          );
        })}
      </select>
      <Button text=">" handleOnClick={() => movePageBy(1)} />
      <Button text=">>" handleOnClick={() => movePageBy(2)} />
    </PaginationStyled>
  );
};

export default Pagination;
