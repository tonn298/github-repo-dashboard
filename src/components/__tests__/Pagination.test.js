import { render, screen, cleanup, fireEvent } from "@testing-library/react";
import Pagination from "../Pagination";
import renderer from "react-test-renderer";
import "@testing-library/jest-dom/extend-expect";

let currentPage = 1;
const pagesArray = [1, 2, 3, 4];
const paginate = (page) => {
  currentPage = page;
};

afterEach(() => {
  cleanup();
});

test("should be able to change between page by dropdown", () => {
  render(
    <Pagination
      paginate={paginate}
      currentPage={currentPage}
      pagesArray={pagesArray}
    />
  );
  expect(screen.getByText("Current page: 1")).toBeVisible();
  fireEvent.click(screen.getByText("Current page: 1"));
  fireEvent.click(screen.getByText("Current page: 2"));
  expect(screen.getByText("Current page: 2")).toBeVisible();
});

test("should be able to change page up and down by 1", () => {
  render(
    <Pagination
      paginate={paginate}
      currentPage={currentPage}
      pagesArray={pagesArray}
    />
  );

  expect(screen.getByText("Current page: 1")).toBeVisible();
  fireEvent.click(screen.getByText(">"));
  expect(screen.getByText("Current page: 2")).toBeVisible();
  fireEvent.click(screen.getByText("<"));
  expect(screen.getByText("Current page: 1")).toBeVisible();
});

test("should be able to change page up and down by 2", () => {
  render(
    <Pagination
      paginate={paginate}
      currentPage={currentPage}
      pagesArray={pagesArray}
    />
  );

  expect(screen.getByText("Current page: 1")).toBeVisible();
  fireEvent.click(screen.getByText(">>"));
  expect(screen.getByText("Current page: 3")).toBeVisible();
  fireEvent.click(screen.getByText("<<"));
  expect(screen.getByText("Current page: 1")).toBeVisible();
});

test("should not be able to go lower than page 1 or higher than page 4", () => {
  render(
    <Pagination
      paginate={paginate}
      currentPage={currentPage}
      pagesArray={pagesArray}
    />
  );

  expect(screen.getByText("Current page: 1")).toBeVisible();
  fireEvent.click(screen.getByText("<"));
  expect(screen.getByText("Current page: 1")).toBeVisible();
  fireEvent.click(screen.getByText(">>"));
  expect(screen.getByText("Current page: 3")).toBeVisible();
  fireEvent.click(screen.getByText(">"));
  expect(screen.getByText("Current page: 4")).toBeVisible();
  fireEvent.click(screen.getByText(">>"));
  expect(screen.getByText("Current page: 4")).toBeVisible();
  fireEvent.click(screen.getByText(">"));
  expect(screen.getByText("Current page: 4")).toBeVisible();
});

test("matches snapshot", () => {
  const tree = renderer.create(
    <Pagination
      paginate={paginate}
      currentPage={currentPage}
      pagesArray={pagesArray}
    />
  );
  expect(tree).toMatchSnapshot();
});
