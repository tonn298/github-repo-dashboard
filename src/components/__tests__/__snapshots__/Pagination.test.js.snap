// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="sc-dlnjwi dpSWgP"
  data-testid="pagination"
>
  <button
    className="sc-bdnxRM hsALWb"
    onClick={[Function]}
  >
    &lt;&lt;
  </button>
  <button
    className="sc-bdnxRM hsALWb"
    onClick={[Function]}
  >
    &lt;
  </button>
  <select
    className="page-dropdown"
    id="page"
    name="page"
    onChange={[Function]}
  >
    <option
      data-testid="option-0"
      value={1}
    >
      Current page: 
      1
    </option>
    <option
      data-testid="option-1"
      value={2}
    >
      Current page: 
      2
    </option>
    <option
      data-testid="option-2"
      value={3}
    >
      Current page: 
      3
    </option>
    <option
      data-testid="option-3"
      value={4}
    >
      Current page: 
      4
    </option>
  </select>
  <button
    className="sc-bdnxRM hsALWb"
    onClick={[Function]}
  >
    &gt;
  </button>
  <button
    className="sc-bdnxRM hsALWb"
    onClick={[Function]}
  >
    &gt;&gt;
  </button>
</div>
`;
